src/imex.rs

```diff
@@ -269,24 +269,47 @@ async fn import_backup(
         context.get_dbfile().display()
     );
 
+    import_backup_stream(context, backup_file, file_size, passphrase).await?;
+    Ok(())
+}
+
+/// Imports backup by reading a tar file from a stream.
+///
+/// `file_size` is ideally the size of the tar file,
+/// but can be an estimate such as the sum of all
+/// archived files without taking headers into account.
+/// It is used to emit progress events.
+pub(crate) async fn import_backup_stream<R: tokio::io::AsyncRead + Unpin>(
+    context: &Context,
+    backup_file: R,
+    file_size: u64,
+    passphrase: String,
+) -> Result<()> {
     let mut archive = Archive::new(backup_file);
 
     let mut entries = archive.entries()?;
     let mut last_progress = 0;
-    while let Some(file) = entries.next().await {
-        let f = &mut file?;
-
+    while let Some(mut f) = entries
+        .try_next()
+        .await
+        .context("Failed to get next entry")?
+    {
         let current_pos = f.raw_file_position();
-        let progress = 1000 * current_pos / file_size;
-        if progress != last_progress && progress > 10 && progress < 1000 {
+        let progress = std::cmp::min(
+            1000 * current_pos.checked_div(file_size).unwrap_or_default(),
+            999,
+        );
+        if progress != last_progress && progress > 10 {
```

Comment:
Minor, but may be shortened to `progress > last_progress` if write `let mut last_progress = 10;` above